pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/RealtimeNativeTextIndex.java

```diff
@@ -0,0 +1,146 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.realtime.impl.invertedindex;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.standard.StandardAnalyzer;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.pinot.segment.local.segment.creator.impl.text.LuceneTextIndexCreator;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableFST;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableFSTImpl;
+import org.apache.pinot.segment.local.utils.nativefst.utils.RealTimeRegexpMatcher;
+import org.apache.pinot.segment.spi.index.mutable.MutableTextIndex;
+import org.roaringbitmap.PeekableIntIterator;
+import org.roaringbitmap.RoaringBitmapWriter;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+public class RealtimeNativeTextIndex implements MutableTextIndex {
+  private final String _column;
+  private final MutableFST _mutableFST;
+  private final RealtimeInvertedIndex _invertedIndex;
+  private final Map<String, Integer> _termToDictIdMapping;
+  private int _nextDocId = 0;
+  private int _nextDictId = 0;
+  private final ReentrantReadWriteLock.ReadLock _readLock;
+  private final ReentrantReadWriteLock.WriteLock _writeLock;
+
+  public RealtimeNativeTextIndex(String column) {
+    _column = column;
+    _mutableFST = new MutableFSTImpl();
+    _termToDictIdMapping = new HashMap<>();
+    _invertedIndex = new RealtimeInvertedIndex();
+
+    ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    _readLock = readWriteLock.readLock();
+    _writeLock = readWriteLock.writeLock();
+  }
+
+  @Override
+  public void add(String document) {
+    List<String> tokens;
+    try {
+      tokens = analyze(document, new StandardAnalyzer(LuceneTextIndexCreator.ENGLISH_STOP_WORDS_SET));
+    } catch (IOException e) {
+      throw new RuntimeException(e.getMessage());
+    }
+
+    try {
+      _writeLock.lock();
+      for (String token : tokens) {
+        Integer currentDictId = _termToDictIdMapping.get(token);
+        if (currentDictId == null) {
+
+          currentDictId = _nextDictId;
```

Comment:
(nit) Can be simplified
```suggestion
          currentDictId = _nextDictId++;
```